x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -0,0 +1,353 @@
+  }
+
+  const possibleOperation = operationDefinition.getPossibleOperationForField(field);
+  if (!possibleOperation) {
+    throw new Error(
+      `Tried to create an invalid operation ${operationDefinition.type} on ${field.name}`
+    );
+  }
+  const isBucketed = Boolean(possibleOperation.isBucketed);
+  if (isBucketed) {
+    return addBucket(layer, operationDefinition.buildColumn({ ...baseOptions, field }), columnId);
+  } else {
+    return addMetric(layer, operationDefinition.buildColumn({ ...baseOptions, field }), columnId);
+  }
+}
+
+export function replaceColumn({
+  layer,
+  columnId,
+  indexPattern,
+  op,
+  field,
+}: {
+  layer: IndexPatternLayer;
+  columnId: string;
+  indexPattern: IndexPattern;
+  op?: OperationType;
+  field?: IndexPatternField;
+}): IndexPatternLayer {
+  const previousColumn = layer.columns[columnId];
+  if (!previousColumn) {
+    throw new Error(`Can't replace column because there is no prior column`);
+  }
+
+  const isNewOperation = Boolean(op) && op !== previousColumn.operationType;
+  const operationDefinition = operationDefinitionMap[op || previousColumn.operationType];
+
+  if (!operationDefinition) {
+    throw new Error('No suitable operation found for given parameters');
+  }
+
+  const baseOptions = {
+    columns: layer.columns,
+    indexPattern,
+    previousColumn: layer.columns[columnId],
+  };
+
+  if (isNewOperation) {
+    // TODO: Reference based operations require more setup to create the references
+
+    if (operationDefinition.input === 'none') {
+      const newColumn = operationDefinition.buildColumn(baseOptions);
+
+      if (previousColumn.customLabel) {
+        newColumn.customLabel = true;
+        newColumn.label = previousColumn.label;
+      }
+
+      return {
+        ...layer,
+        columns: adjustColumnReferencesForChangedColumn(
+          { ...layer.columns, [columnId]: newColumn },
+          columnId
+        ),
+      };
+    }
+
+    if (!field) {
+      throw new Error(`Invariant error: ${operationDefinition.type} operation requires field`);
+    }
+
+    const newColumn = operationDefinition.buildColumn({ ...baseOptions, field });
+
+    if (previousColumn.customLabel) {
+      newColumn.customLabel = true;
+      newColumn.label = previousColumn.label;
+    }
+
+    const newColumns = { ...layer.columns, [columnId]: newColumn };
+    return {
+      ...layer,
+      columnOrder: getColumnOrder({ ...layer, columns: newColumns }),
+      columns: adjustColumnReferencesForChangedColumn(newColumns, columnId),
+    };
+  } else if (
+    operationDefinition.input === 'field' &&
+    field &&
+    ('sourceField' in previousColumn ? field.name !== previousColumn.sourceField : true)
+  ) {
+    const newColumn = operationDefinition.buildColumn({ ...baseOptions, field, previousColumn });
+
+    // TODO: This is probably wrong, not all params are shared on all ops?
+    if ('params' in previousColumn) {
+      newColumn.params = previousColumn.params;
```

Comment:
Where is this logic coming from? I think in the old implementation this was handled by https://github.com/elastic/kibana/pull/82550/files#diff-152b44f262dc52b3cbd14022b79c48972bfbffe04c1d2d4827785b781f5fdd89L80-L86

which seems like the right approach to me - re-use the params if the operation type doesn't change, otherwise leave it up to the buildColumn implementation of the target operation definition to handle it.

Reply:
We already have a bunch of this logic duplicated, especially in `buildColumn` functions. I think this is where I need help doing some manual testing because we don't have automated tests around this.